Ignore deprecated progress event for blocks with new Completion API.

tomaszgy opened the pull request:
This PR changes the new completion API to ignore progress events from XBlocks that declare `has_custom_completion`. That way, XBlocks can emit both events without causing duplicate records.

The rationale behind is to allow a single codebase for each XBlock to work with the old solutions fork progress code and the new upstream code simultaneously; to make the upgrade of solutions to the new completion API smoother. For the details, see [this GitHub comment](https://github.com/edx-solutions/xblock-ooyala/pull/78#issuecomment-346491707).

**JIRA tickets**: Implements [OC-3393](https://tasks.opencraft.com/browse/OC-3393).

**Dependencies**: None

**Sandbox URL**: TBD

**Merge deadline**: None

**Testing instructions**:

1. Please run added tests and see them pass.

**Reviewers**
- [ ] @bradenmacdonald / @jcdyer 
- [ ] edX reviewer[s] TBD

**Extra**

Rebased from fe266ddc5ade491376bae157dc79672730ee97be onto master.
Squashed from 635d4e7429a34a6b6448fe86e960a460a1ca73f0 and then rebased onto master. 

Signed-off-by: Tomasz Gargas <tomasz@opencraft.com>

edx-webhook commented:
Thanks for the pull request, @tomaszgy! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=16741"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

bradenmacdonald commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -554,12 +554,14 @@ def handle_deprecated_progress_event(block, event):
-                completion=1.0,
-            )
+
+            if not hasattr(block, 'has_custom_completion'):
```
Should be `if not getattr(block, `has_custom_completion`, False):` otherwise this can't handle blocks with `has_custom_completion = False`

bradenmacdonald replied:
Also, please add a comment to explain that if blocks explicitly declare support for the new completion API, we expect them to emit 'completion' events, and we ignore the deprecated 'progress' events in order to avoid duplicate work and support slightly different semantics.

tomaszgy replied:
👍

jcdyer commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -554,12 +554,18 @@ def handle_deprecated_progress_event(block, event):
+            # If blocks explicitly declare support for the new completion API,
+            # we expect them to emit 'completion' events,
+            # and we ignore the deprecated 'progress' events
+            # in order to avoid duplicate work and support slightly different semantics.
```
Nit: support doesn't quite fit grammatically.  I would say "in order to avoid duplicate work and possibly conflicting semantics."

tomaszgy replied:
👍

tomaszgy commented:
@jcdyer thanks for your review! I've addressed your comment.

tomaszgy commented:
@robrap, would you be able to review this PR? OpenCraft's review is finished already. Thanks!

robrap commented:
@tomaszgy: I am the right person for these for now.  I've added it to my list and should get to it in the next couple of days.  I'm just wrapping up another piece of work.  Thank you.

tomaszgy commented:
@robrap, thanks! Looking forward to seeing your review :)

iloveagent57 commented on lms/djangoapps/courseware/tests/test_module_render.py:
```diff
@@ -636,32 +653,70 @@ def test_completion_events_with_completion_disabled(self, signal, data):
+
+    @XBlock.register_temp_plugin(StubCompletableXBlock, identifier='comp')
+    def test_progress_signal_ignored_for_completable_xblock(self):
+        with completion_waffle.waffle().override(completion_waffle.ENABLE_COMPLETION_TRACKING, True):
```
Minor thing: lines 680-695 and lines 701-716 are identical with the exception of the value of `category` - so it might be worth extracting a little helper method that returns `response`, have the test method use that, and then make assertions.

iloveagent57 commented on lms/djangoapps/courseware/tests/test_module_render.py:
```diff
@@ -143,6 +144,22 @@ def progress(self, json_data, suffix):  # pylint: disable=unused-argument
         return self.runtime.publish(self, 'progress', {})
 
 
+class XBlockWithoutCompletionAPI(XBlock):
```
Is there a test to cover the case of XBlocks that do emit the 'progress' signal and have `has_custom_completion = True`?  That is, something to cover the `handle_deprecated_progress_event()` function's `else` path that doesn't cause a `BlockCompletion` record to be created?

tomaszgy replied:
Yes, there is one: [`test_progress_signal_ignored_for_completable_xblock`](https://github.com/open-craft/edx-platform/blob/2c70e38e7d9b92e6d5b374611d62d90cc7c61a50/lms/djangoapps/courseware/tests/test_module_render.py#L679).

iloveagent57 replied:
Ah, sorry, I must have been conflating two tests in my mind with my earlier comment.

iloveagent57 commented on lms/djangoapps/courseware/tests/test_module_render.py:
```diff
@@ -636,32 +653,70 @@ def test_completion_events_with_completion_disabled(self, signal, data):
+            completion = BlockCompletion.objects.get(block_key=block.scope_ids.usage_id)
+            self.assertEqual(completion.completion, 0.625)
+
+    @XBlock.register_temp_plugin(StubCompletableXBlock, identifier='comp')
```
(related to my comment above)
Should this actually be a `XBlockWithoutCompletionAPI` XBlock?

iloveagent57 commented:
@tomaszgy or @jcdyer bump

tomaszgy commented:
@iloveagent57 Thanks for a review and a ping! Sorry about the delay, I'll get to this soon (if possible before Christmas).

tomaszgy commented:
@iloveagent57 Thanks again for your review! Just addressed your comments (please see one new commit and a comment I've left responding to one of yours).

tomaszgy commented:
@iloveagent57, great, thank you! I've squashed the commits and rebased this branch onto current master so now the Python unit tests should pass and this should be ready to merge.

bradenmacdonald commented:
@iloveagent57 Thanks for reviewing this! Since @jcdyer is off today, would you be able to merge this for us? (Or I can but I didn't review it myself so would prefer one of the reviewers to merge.)

tomaszgy commented:
@iloveagent57, a friendly ping :)

bradenmacdonald commented:
@tomaszgy Do you mind rebasing this on master one more time to reduce risk, and then ping me and @jcdyer, and one of us will merge this for you?

tomaszgy commented:
@bradenmacdonald, sure, just rebased it.
CC: @jcdyer
